### DISCUSSION TITLE
model.predict for graph classification

### DISCUSSION THREAD

**Discussion by Digital-Chemist** [danielegrattarola/spektral#369] [Q&A]
Not exactly sure how to use model.predict() for graph classification.  I've trained a GeneralGNN model with some custom data using the framework in the example with about 130 graphs (https://github.com/danielegrattarola/spektral/blob/master/examples/graph_prediction/general_gnn.py).  No issues with the training/testing.  If I then have a single unlabeled Graph (no y label) created the exact same way that the training/testing set graphs were created except no y=label, should I not just be able to feed that into model.predict as 

```
graphX = Graph.x
graphA = Graph.a
predict = model.predict([graphX, graphA])
```
This gives an error:

`global_pool.py", line 30, in call  *
                I = inputs[1]
IndexError: list index out of range`

If I put the single Graph into the same loader (DisjointLoader) created in the same way that the training/testing set was created, I get an error:

```
general_gnn.py", line 163, in call  *
            x, a = inputs
        OperatorNotAllowedInGraphError: Iterating over a symbolic `tf.Tensor` is not allowed: AutoGraph did convert this function. This might indicate you are trying to use an unsupported feature.
    Call arguments received by layer "general_gnn_1" (type GeneralGNN):
      • inputs=tf.Tensor(shape=(None, None), dtype=float32) 
```
Both x and a are definitely in the single dataset (print(pred_dataset[0].x) and print(pred_dataset[0].a) give the expected arrays and sizes.  

What am I missing here?  Any help would be greatly appreciated.

**Comment by danielegrattarola** [comment]
Hi, 

that example is using DisjointLoader and graph-level predictions, so the model expects a batch index input (typically called `i` in the documentation of the library) to indicate which nodes belong to which graphs. In this case, you only have one graph so the input is just an array of zeros with shape `(x.shape[0], )`

Change the call to 

```py
model.predict([graphX, graphA, tf.zeros(x.shape[0])])
```

**Reply by Digital-Chemist** [reply]
Thank you very much for the lead.  For reference, actually had to do:

`model.predict([graphX, graphA, tf.zeros(graphX.shape[0], dtype=tf.dtypes.int32)])`

as the input needs to be int32 or int64 or else it errors out at global sum pooling.  May want to include a model prediction section in the documentation for cases like this.  Especially because you can't exactly search the documentation for "i".  

Thanks again.

**Reply by danielegrattarola** [reply]
Yeah I will probably change the `x,a,i` nomenclature sometime soon.

Glad you solved the problem!
